rust/agents/relayer/src/msg/metadata_builder.rs

```diff
@@ -1,45 +1,85 @@
+use hyperlane_core::{HyperlaneMessage, Signers};
 use hyperlane_core::{Mailbox, MultisigIsm, H256};
+use tokio::sync::RwLock;
+use tracing::{info, instrument};
+
+use crate::merkle_tree_builder::MerkleTreeBuilder;
 
 #[derive(Debug, Clone)]
 pub struct MetadataBuilder {
     metrics: Arc<CoreMetrics>,
     signer: Option<Signers>,
     chain_setup: ChainSetup,
+    checkpoint_syncer: MultisigCheckpointSyncer,
+    prover_sync: Arc<RwLock<MerkleTreeBuilder>>,
 }
 
 impl MetadataBuilder {
     pub fn new(
         metrics: Arc<CoreMetrics>,
         signer: Option<Signers>,
         chain_setup: ChainSetup,
+        checkpoint_syncer: MultisigCheckpointSyncer,
+        prover_sync: Arc<RwLock<MerkleTreeBuilder>>,
     ) -> Self {
         MetadataBuilder {
             metrics,
             signer,
             chain_setup,
+            checkpoint_syncer,
+            prover_sync,
         }
     }
 
+    #[instrument(err, skip_all, fields(msg_nonce=message.nonce))]
     pub async fn fetch_metadata(
         &self,
         message: HyperlaneMessage,
         mailbox: CachingMailbox,
-        checkpoint: MultisigSignedCheckpoint,
-        proof: Proof,
     ) -> eyre::Result<Vec<u8>> {
         let ism_address = mailbox.recipient_ism(message.recipient).await?;
         let multisig_ism = self.build_multisig_ism(ism_address).await?;
-
-        let metadata = multisig_ism
-            .format_metadata(message.clone(), &checkpoint, proof)
+        let validators_and_threshold = multisig_ism
+            .validators_and_threshold(message.clone())
             .await?;
-        Ok(metadata)
+        let validators = validators_and_threshold.0;
+        // TODO: we don't want to fetch checkpoints that are more recent than what we can create a proof against?
+        if let Some(checkpoint) = self
+            .checkpoint_syncer
+            .fetch_checkpoint_in_range(
+                validators.clone(),
+                validators_and_threshold.1.into(),
+                self.prover_sync.read().await.count() - 1,
+                message.nonce,
+            )
+            .await?
+        {
+            info!(
+                checkpoint_index = checkpoint.checkpoint.index,
+                signature_count = checkpoint.signatures.len(),
+                "Found checkpoint with quorum"
+            );
+
+            let proof = self
+                .prover_sync
+                .read()
+                .await
+                .get_proof(message.nonce, checkpoint.checkpoint.index)?;
+            assert_eq!(checkpoint.checkpoint.root, proof.root());
+            let metadata = multisig_ism.format_metadata(
+                validators.clone(),
+                validators_and_threshold.1,
+                &checkpoint,
+                proof,
+            );
+            Ok(metadata)
+        } else {
+            // TODO: Figure out how to do proper error reporting. Should probably have the checkpoint syncer return errors rather than an option
```

Comment:
Do this

Reply:
Actually, looking for opinions here.

Sovereign consensus will expand the surface area where improper user config can cause the relayer to run into issues that look like errors but are actually proper (relayer) behavior. e.g.:
- User ISM does not specify the proper interface (e.g. `multisig_ism.validatorsAndThreshold()` fails)
- The relayer isn't aware of the validator checkpoints needed to submit the message

We have this situation already in:
- Validators may not have signed a checkpoint matching their `latest_checkpoint.json` contents. I believe we handle this by returning `None`
- Validators may have signed a checkpoint that doesn't match the actual root. I believe here we return a `MerkleTreeBuilderError::MismatchedRoots`
- A message may not be deliverable due to the recipient rejecting it. Here we return `Ok(false)`

Wondering if we should try to standardize on what to do if something unexpected happens that is outside of the relayer's domain.